### PR TITLE
Switch to svd-rs and svd-encoder

### PR DESCRIPTION
**Pull request by jeandudey**
Will wait on resolution of #32 and #33 to rebase any changes if necessary.

Some behavior is different:

- `width` is specified now at `/device` level. Required by schema. A command line flag would be good for this and/or the address size.
- `version` is also mandatory and specified.
- `size` field is now specified in hex rather than decimal, this is consistent with most SVD files. It's not handled directly by us but by `svd-rs`. E.g `<size>32</size>` is now `<size>0x20</size>`.
- Interrupts are matched with their peripheral and thus a `CPU` peripheral is not required per se. Interrupts in ATDF files include the module name so they are attached correctly.

### CODE REVIEW

File: src/atdf/peripheral.rs [hunk 1]
```diff
@@ -45,7 +46,7 @@ pub fn parse_list(
                     })
                     .filter(|e| e.name == "register")
                 {
-                    registers.push(atdf::register::parse(register, offset, &value_groups)?);
+                    registers.push(atdf::register::parse(register, offset.try_into().unwrap(), &value_groups)?);
```
Comment (jeandudey) [inline src/atdf/peripheral.rs:49]: Just a self note to remove that unwrap

File: src/atdf/register.rs [hunk 2]
```diff
@@ -40,7 +40,7 @@ pub fn parse(
     Ok(chip::Register {
         name,
         description,
-        address: util::parse_int(el.attr("offset")?)? + offset,
+        address: (util::parse_int(el.attr("offset")?)? + offset),
```
Comment (jeandudey) [inline src/atdf/register.rs:43]: Unrelated change o.O

**Comment by jeandudey** [comment]
Investigated in the atmega atdf files but it's inconsistent, some of them do specify the module-instance some don't. I'll switch to the old behavior, I don't have an urgent need yet for interrupts number so I think it's fine the way it is currently done. Maybe later I can work on a more thorough solution that takes both situations into account.

**Comment by Rahix** [comment]
Yeah, there is unfortunately quite a lot of variance in the different ATDF files.  I should really look into adding a regression suite and CI in general for this repository.

Both of your other PRs are now merged so please rebase this one :)

**Comment by jeandudey** [comment]
Re-added back the default register properties and the previous interrupt behavior. I think it generates more or less the same SVD files with the exceptions listed in the description and the placement inside the file itself, same content just differently positioned.

File: src/svd/register.rs [hunk 3]
```diff
@@ -1,48 +1,39 @@
...
+use crate::svd::restriction::generate_access;
+
+pub fn generate(r: &chip::Register, base: u32) -> crate::Result<svd_rs::Register> {
+    let (write_constraint, _) = crate::svd::restriction::generate(&r.restriction, r.size as u32 * 8)?;
+
+    let register = svd_rs::RegisterInfo::builder()
+        .name(r.name.clone())
+        .description(r.description.clone().or_else(|| {
+            log::warn!("Description missing for register \"{}\"", r.name);
+            None
```
[the remark this answers is not in the document]
Comment (Rahix) [reply src/svd/register.rs:11]: Here's a diff for you:

```diff
diff --git a/src/svd/field.rs b/src/svd/field.rs
index 7210f63..e5fa0ec 100644
--- a/src/svd/field.rs
+++ b/src/svd/field.rs
@@ -11,7 +11,7 @@ pub fn generate(f: &chip::Field) -> crate::Result<svd_rs::Field> {
         .name(f.name.clone())
         .description(f.description.clone().or_else(|| {
             log::warn!("Description missing for field {:?}", f.name);
-            Some("No description.".to_owned())
+            Some("No Description.".to_owned())
         }))
         .bit_range(svd_rs::BitRange {
             offset: lsb,
diff --git a/src/svd/interrupt.rs b/src/svd/interrupt.rs
index 4052ccf..9e3ca0b 100644
--- a/src/svd/interrupt.rs
+++ b/src/svd/interrupt.rs
@@ -14,7 +14,10 @@ pub fn generate(peripherals: &mut [svd_rs::Peripheral], c: &chip::Chip) -> crate
                 .map(|interrupt| {
                     svd_rs::Interrupt::builder()
                         .name(interrupt.name.clone())
-                        .description(interrupt.description.clone())
+                        .description(interrupt.description.clone().or_else(|| {
+                            log::warn!("Description missing for interrupt {:?}", interrupt.name);
+                            Some("No Description.".to_owned())
+                        }))
                         .value(interrupt.index.try_into().map_err(crate::Error::from)?)
                         .build(svd_rs::ValidateLevel::Strict)
                         .map_err(crate::Error::from)
diff --git a/src/svd/peripheral.rs b/src/svd/peripheral.rs
index 71bf691..241c4f5 100644
--- a/src/svd/peripheral.rs
+++ b/src/svd/peripheral.rs
@@ -61,7 +61,7 @@ pub fn generate(p: &chip::Peripheral) -> crate::Result<svd_rs::Peripheral> {
         .name(p.name.clone())
         .description(p.description.clone().or_else(|| {
             log::warn!("Description missing for peripheral {:?}", p.name);
-            None
+            Some("No Description.".to_owned())
         }))
         .base_address(u64::from(base))
         .address_block(create_address_blocks(p)?)
diff --git a/src/svd/register.rs b/src/svd/register.rs
index 5cbca6f..fd08ca8 100644
--- a/src/svd/register.rs
+++ b/src/svd/register.rs
@@ -8,7 +8,7 @@ pub fn generate(r: &chip::Register, base: u32) -> crate::Result<svd_rs::Register
         .name(r.name.clone())
         .description(r.description.clone().or_else(|| {
             log::warn!("Description missing for register \"{}\"", r.name);
-            None
+            Some("No Description.".to_owned())
         }))
         .address_offset(r.address as u32 - base)
         .size(if r.size != 0 {
diff --git a/src/svd/restriction.rs b/src/svd/restriction.rs
index e7b7f31..42b807f 100644
--- a/src/svd/restriction.rs
+++ b/src/svd/restriction.rs
@@ -63,7 +63,7 @@ pub fn generate_enumerated(e: &chip::EnumeratedValue) -> crate::Result<svd_rs::E
         .name(e.name.clone())
         .description(e.description.clone().or_else(|| {
             log::warn!("Description missing for enumeratedValue {:?}", e.name);
-            None
+            Some("No Description.".to_owned())
         }))
         .value(Some(e.value.try_into()?))
         .build(svd_rs::ValidateLevel::Strict)
```

between this and the other comment, I get zero changes in the generated code on my end so this is looking _very_ good!

**Review by Rahix — COMMENTED** [review]
Alright, I ran tests with this branch now.  Things look very promising, only two topics left from my point of view.  See comments below.

Thanks again for your work here!

File: src/svd/restriction.rs [hunk 4]
```diff
@@ -1,79 +1,80 @@
...
-                })
                 .map(generate_enumerated)
-                .collect::<Result<_, _>>()?;
+                .collect::<Result<Vec<_>, _>>()?;
 
-            vec![wc, values_el]
+            let enumerated_values = svd_rs::EnumeratedValues::builder()
+                .values(values)
+                .build(svd_rs::ValidateLevel::Strict)
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (jeandudey) [reply src/svd/restriction.rs:48]: I agree with the approach, the unwrap was probably a left over since I did all the code using unwrap first then added error handling. I'll change that.

File: src/svd/register.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeandudey) [reply src/svd/register.rs:11]: I'm fine with dummy descriptions, just thought of removing them as they aren't strictly necessary per the schema.

[the remark this answers is not in the document]
Comment (jeandudey) [reply src/svd/register.rs]: Added descriptions back on f381ae10da9f239584282ff954164e16c0993fff

File: src/svd/restriction.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeandudey) [reply src/svd/restriction.rs]: Solved, also removed another unwrap that I left on the encode function.

**Review by Rahix — APPROVED** [review]
With this an #35 applied, I get zero regressions.  Thanks a lot for your work on this conversion!
